File: nodejs-wdio/tests/pages/components/signInForm.comp.ts
import Page from "../page";
import { User } from "../../types/user";

class SignInComponent extends Page {
  constructor() {
    super("");
  }
  protected get loginField() {
    return $("[placeholder='you@example.com']");
  }

  protected get pwdField() {
    return $("[placeholder='Enter password']");
  }

  protected get submitButton() {
    return $(".submit-button");
  }

  protected get fieldErrors() {
    return $$(".base-form-error");
  }

  protected get formError() {
    return $(".form-error-message");
  }

  public signIn(user: User): void {
    this.allureReporter.startStep("Sign In");
    this.setEmail(user.userEmail);
    this.setPassword(user.userPwd);
    this.submit();
    this.allureReporter.endStep();
  }

  public isSignedIn(): boolean {
    return false;
  }

  public signInAndReturnFieldsError(user: User): Array<String> {
    this.signIn(user);
    this.allureReporter.startStep("Get Fields Error Messages");
    browser.waitUntil(() => this.fieldErrors.some((e) => e.isDisplayed()));
    var errorMessages = this.fieldErrors.reduce<string[]>(function (
      filtered,
      e
    ) {
      if (e.isDisplayed()) filtered.push(e.getText());
      return filtered;
    },
    []);
    this.allureReporter.endStep();
    return errorMessages;
  }

  public signInAndReturnFormError(user: User): string {
    this.signIn(user);
    this.allureReporter.startStep("Get Form Error Message");
    browser.waitUntil(() => this.formError.isDisplayed() === true);
    const errorMessage = this.formError.getText();
    this.allureReporter.endStep();
    return errorMessage;
  }

  private setEmail(email: string): void {
    this.allureReporter.startStep("Set Email");
    this.loginField.setValue(email);
    this.allureReporter.endStep();
  }

  private setPassword(password: string): void {
    this.allureReporter.startStep("Set Password");
    this.pwdField.setValue(password);
    this.allureReporter.endStep();
  }

  private submit(): void {
    this.allureReporter.startStep("Submit");
    this.submitButton.click();
    this.allureReporter.endStep();
  }
}

export default new SignInComponent();

File: ruby-capybara/tests/pages/components/sign_in_form_comp.rb
require_relative '../base_page'

class SignInFormComponent < BasePage
  SCREEN_SELECTOR = ''

  LOGIN_FIELD = 'you@example.com'
  PASSWORD_FIELD = 'Enter password'
  SUBMIT_BUTTON  = 'Ready to Livestorm'
  FIELDS_ERROR = '.base-form-error'
  FORM_ERROR = '.form-error-message'

  def initialize()
    super(SCREEN_SELECTOR)
  end

  def sign_in(user)
    fill_email(user['userEmail'])
    fill_password(user['userPwd'])
    submit
  end

  def sign_in_and_return_fields_error(user)
    sign_in(user)
    return page.all(FIELDS_ERROR, :visible => true).map { |el| el.text}
  end

  def sign_in_and_return_form_error(user)
    sign_in(user)
    return find(FORM_ERROR, :visible => true).text
  end

  private

  def fill_email(email)
    fill_in LOGIN_FIELD, with: email
  end

  def fill_password(password)
    fill_in PASSWORD_FIELD, with: password
  end

  def submit
    click_button SUBMIT_BUTTON
  end
end
